src/Space.ts

```diff
@@ -1,11 +1,116 @@
 
-class Space {
+const ERROR_CLIENT_ALREADY_ENTERED = 'Client has already entered the space';
 
-  constructor(
-    private spaces: Spaces
-  ){}
+class MemberUpdateEvent extends Event {
+  constructor(public members: SpaceMember[]) {
+    super('memberUpdate', {});
+  }
 }
 
+const createSpaceMemberFromPresenceMember = (m: Types.PresenceMessage): SpaceMember => ({
+  clientId: m.clientId as string,
+  isConnected: true,
+  data: JSON.parse(m.data as string),
+});
+
+class Space extends EventTarget {
+  private members: SpaceMember[] = [];
+  private channelName: string;
+  private channel: Types.RealtimeChannelPromise;
+
+  eventTarget: EventTarget;
+
+  constructor(private name: string, private client: Types.RealtimePromise, private options?: SpaceOptions) {
+    super();
+    this.setChannel(this.name);
+  }
+
+  private setChannel(rootName) {
+    // The channel name prefix here should be unique to avoid conflicts with non-space channels
+    this.channelName = `_ably_space_${rootName}`;
+    this.channel = this.client.channels.get(this.channelName);
+  }
+
+  private createMember(clientId: string, isConnected: boolean, data: { [key: string]: any }) {
+    this.members.push({ clientId, isConnected, data });
+  }
+
+  private syncMembers() {
+    this.channel.presence.get().then((presenceMessages) => {
+      this.members = presenceMessages.filter((m) => m.clientId).map(createSpaceMemberFromPresenceMember);
+    });
+  }
+
+  private subscribeToPresenceEvents() {
+    this.channel.presence.subscribe('enter', (message: Types.PresenceMessage) => {
+      this.updateMemberState(message.clientId, true, JSON.parse(message.data as string));
+    });
+
+    this.channel.presence.subscribe('leave', (message: Types.PresenceMessage) => {
+      this.updateMemberState(message.clientId, false);
+    });
+
+    this.channel.presence.subscribe('update', (message: Types.PresenceMessage) => {
+      this.updateMemberState(message.clientId, true, JSON.parse(message.data as string));
+    });
+  }
+
+  private updateMemberState(clientId: string | undefined, isConnected: boolean, data?: { [key: string]: any }) {
+    const implicitClientId = clientId ?? this.client.auth.clientId;
+
+    if (!implicitClientId) {
+      return;
+    }
+
+    const member = this.members.find((m) => m.clientId === clientId);
+
+    if (!member) {
+      this.createMember(implicitClientId, isConnected, data || {});
+    } else {
+      member.isConnected = isConnected;
+      if (data) {
+        // Member data is completely overridden, except lastEventTimestamp which is updated
+        member.data = {
+          ...data,
+          lastEventTimestamp: new Date(),
+        };
+      }
+    }
+
+    const memberUpdateEvent = new MemberUpdateEvent(this.members);
+    this.dispatchEvent(memberUpdateEvent);
+  }
+
+  enter(data: unknown) {
+    if (!data || typeof data !== 'object') {
+      return;
+    }
+
+    const clientId = this.client.auth.clientId || undefined;
```

Comment:
When might this be undefined? Would it make sense for there to be a step in the initialisation/connection sequence where the client discovers its own `clientId` if it didn't know it already?

Reply:
This is really just to cater for the ably-js SDK TypeScript typings. Currently, the same clientId value (`string || undefined`) is used both for points where the clientId really might be undefined (when initializing the Ably client) and for every other point in the codebase, where it can't be undefined.
In the Ably JS SDK, this could be solved by simply having clientId be a `string` and let it be optional in the one place it was possible to be undefined.

Reply:
@owenpearson ^ worth raising an issue?